wordpress-mobile/AztecEditor-Android
# Add UI Tests for Simple Text Formatting

This PR adds a suite of UI tests for simple text formatting in the Aztec demo app, using the Android Espresso testing framework.

## Before running the tests

Espresso [advises](https://google.github.io/android-testing-support-library/docs/espresso/setup/#setup-your-test-environment) disabling system animations on devices used for testing:

> On your device, under Settings->Developer options disable the following 3 settings:
> 
> - Window animation scale
> - Transition animation scale
> - Animator duration scale

One additional setup step is also required to handle an Espresso issue with clicks (see the caveats below):

On your device, under Settings -> Accessibility -> Touch & hold delay, set the delay to `Long`.

## Test Details

Each test performs the following steps:

1. Enter text (unformatted).
2. Select a formatting button in the toolbar.
3. Enter more text (with formatting).
4. Verify the correct HTML appears in HTML mode for the unformatted and formatted text.

A couple caveats:

- The Espresso framework sometimes performs a long click instead of a single click (see the [Espresso code](https://android.googlesource.com/platform/frameworks/testing/+/android-support-test/espresso/core/src/main/java/android/support/test/espresso/action/GeneralClickAction.java#75) for an explanation). In the Aztec toolbar, this displays a hint instead of selecting the formatting button and causes the test to fail.
- Espresso also has trouble with the editor's link dialog. Although it can identify the fields and buttons in the dialog (and TalkBack works with the dialog as expected) it clicks in the wrong position. This seems to be a bug with the coordinates select for the click action for a custom `AlertDialog` and so I reversed the link formatting test. (It enters the link HTML and then verifies that the link dialog displays the expected values.)

cc @catehstn @lancewillett @hoverduck
